src/libexpr/eval.cc

```diff
@@ -701,21 +701,22 @@ Path EvalState::toRealPath(const Path & path, const NixStringContext & context)
 }
 
 
-Value * EvalState::addConstant(const std::string & name, Value & v)
+Value * EvalState::addConstant(const std::string & name, Value & v, Constant info)
 {
     Value * v2 = allocValue();
     *v2 = v;
-    addConstant(name, v2);
+    addConstant(name, v2, info);
     return v2;
 }
 
 
-void EvalState::addConstant(const std::string & name, Value * v)
+void EvalState::addConstant(const std::string & name, Value * v, Constant info)
 {
     staticBaseEnv->vars.emplace_back(symbols.create(name), baseEnvDispl);
     baseEnv.values[baseEnvDispl++] = v;
     auto name2 = name.substr(0, 2) == "__" ? name.substr(2) : name;
     baseEnv.values[0]->attrs->push_back(Attr(symbols.create(name2), v));
+    constantInfos.push_back({name2, info});
```

Comment:
@roberth because we are using `name2`, there is no `__` appearing in the docs.

I do think this `__` stripping in the implementation is very crude, and it would be better to have some sort of `bool exposed;` field that would *prepend* `__` as needed, but I figured that was better as a follow-up pure refactor.

(I would also want to do it consistently for primitive operations and primitive constants. Which will be easier after this PR too.)

Reply:
(Speaking of unifying things, I also think the difference between a builtin constant and a builtin primitive operation is an implementation detail we should not be exposing in the documentation to users. `derivation` being a function that is a builtin-in constant rather than primitive operation is the operative example. But that is also best as a follow-up.)